Orange/tests/test_statistics.py

```diff
@@ -241,3 +228,94 @@ def test_var(self):
                     var(csr_matrix(data), axis=axis),
                     np.var(data, axis=axis)
                 )
+
+
+class TestCountnans(unittest.TestCase):
+    def test_countnans_on_dense_array(self):
+        x = np.array([0, 1, 0, 2, 2, np.nan, 1, np.nan, 0, 1])
+        self.assertEqual(countnans(x), 2)
+
+    def test_countnans_on_sparse_array(self):
+        x = csr_matrix([0, 1, 0, 2, 2, np.nan, 1, np.nan, 0, 1])
+        self.assertEqual(countnans(x), 2)
+
+    def test_shape_matches_dense_and_sparse_given_array_and_axis_None(self):
+        dense = np.array([0, 1, 0, 2, 2, np.nan, 1, np.nan, 0, 1])
+        sparse = csr_matrix(dense)
+        np.testing.assert_equal(countnans(dense), countnans(sparse))
+        self.assertEqual(countnans(dense), 2)
+
+    def test_shape_matches_dense_and_sparse_given_array_and_axis_0(self):
+        dense = np.array([0, 1, 0, 2, 2, np.nan, 1, np.nan, 0, 1])
+        sparse = csr_matrix(dense)
+        np.testing.assert_equal(countnans(dense, axis=0), countnans(sparse, axis=0))
+        self.assertEqual(countnans(dense, axis=0), 2)
+
+    def test_shape_matches_dense_and_sparse_given_array_and_axis_1(self):
+        dense = np.array([0, 1, 0, 2, 2, np.nan, 1, np.nan, 0, 1])
+        sparse = csr_matrix(dense)
+        np.testing.assert_equal(countnans(dense, axis=1), countnans(sparse, axis=1))
+        self.assertEqual(countnans(dense, axis=1), 2)
+
+    def test_countnans_on_2d_matrix(self):
+        x = [[1, np.nan],
+             [2, np.nan]]
+        np.testing.assert_equal(
+            countnans(x), 2, 'Countnans fails on dense data')
+        np.testing.assert_equal(
+            countnans(csr_matrix(x)), 2, 'Countnans fails on sparse data.')
+
+    def test_countnans_on_columns(self):
+        x = [[1, np.nan],
+             [2, np.nan]]
+        np.testing.assert_equal(
+            countnans(x, axis=0), [0, 2],
+            'Countnans fails on dense data with `axis=0`')
+        np.testing.assert_equal(
+            countnans(csr_matrix(x), axis=0), [0, 2],
+            'Countnans fails on sparse data with `axis=0`')
+
+    def test_countnans_on_rows(self):
+        x = [[1, np.nan],
+             [2, np.nan]]
+        np.testing.assert_equal(
+            countnans(x, axis=1), [1, 1],
+            'Countnans fails on dense data with `axis=1`')
+        np.testing.assert_equal(
+            countnans(csr_matrix(x), axis=1), [1, 1],
+            'Countnans fails on sparse data with `axis=1`')
+
+    def test_countnans_with_2d_weights(self):
```

Comment:
This didn't work before and I wasn't sure whether or not it was even worth the hassle, since I assume we only use sparse matrices when the data is very big, and forming a dense 2d weight matrix of that size kind of defeats the purpose...

However, I've implemented it now. In any case sparse weight matrices weren't supported before and aren't supported now. I don't know if it's a requirement, but it would require more work.

Reply:
You're right, we shouldn't worry about weights on sparse matrices.